### PR TITLE
Note changes in bytecode between versions

### PR DESCRIPTION
**Pull request by rocky**
I'm done with a first pass. Since this is large, short of whatever small things we can catch, I'd like this merged in even though there are gaps in filling out newly added instructions, and other things of the "finish completing..." nature. For that, I think we just make more passes.

I'll be branching off of this branch for the other couple of ideas I have. (And that's another reason why I'd like this back into master - I'd rather branch from there.)

### CODE REVIEW

File: elisp-bytecode.texi [hunk 1]
```diff
@@ -537,7 +538,38 @@ opcode, in little-endian byte order.
...
-A stack reference
+A stack reference. This is used only in lexical binding.
+
+Added in Emacs 24.1
+
+@subsubsection Warning
+Running an instruction with opcode 0 (logically this would be called
+@code{stack-ref[0]}), will cause an immediate abort of Emacs!
+
+0 is typically an invalid bytecode and assembly instructions because 0
```
Comment (pipcet) [inline elisp-bytecode.texi:549]: It's an invalid opcode, but a common and valid byte in a bytecode string. I think that's worth pointing out.

File: elisp-bytecode.texi [hunk 2]
```diff
@@ -1029,7 +1093,10 @@ Call @code{+} with two arguments, adding the two values at the top of the stack.
 @node mult
 @unnumberedsubsubsec @code{mult} (95)
 @kindex mult
-Call @code{*} with two arguments, multiplying the two values at the top of the stack.
+Call @code{*} with two arguments, multiplying the two values at the
+top of the stack.
+
+Since emacs 19.34
```
Comment (pipcet) [inline elisp-bytecode.texi]: bytecomp.el incorrectly states "v19 only" for this opcode, but it's still emitted.

File: elisp-bytecode.texi [hunk 3]
```diff
@@ -1063,6 +1132,7 @@ function for them.
 * concat2::
 * concat3::
 * concat4::
+* concatN::
```
Comment (pipcet) [inline elisp-bytecode.texi:1157]: I had deliberately excluded concatN and listN from this section, but I think it's better to document them here rather than having a special section for them, now.

File: elisp-bytecode.texi [hunk 4]
```diff
@@ -1090,26 +1160,44 @@ Call @code{concat} with three arguments.
 Call @code{concat} with four arguments.
 
 
+@node concatN
+@unnumberedsubsubsec @code{concatN} (174)
+@kindex concatN
+
+Call @code{concat} an arbitrary number of arguments.  Note there are
```
Comment (pipcet) [inline elisp-bytecode.texi]: in the 0-255 range, concat with more arguments will generate a function call.

File: elisp-bytecode.texi [hunk 5]
```diff
@@ -1224,76 +1323,128 @@ Call @code{current-buffer} with no arguments.
 @kindex set-buffer
 Call @code{set-buffer} with one argument.
 
+@node save-current-buffer-1
```
Comment (pipcet) [inline elisp-bytecode.texi]: I don't think we should call this "save-current-buffer-1", since "byte-save-current-buffer" is what bytecomp.el uses.

File: elisp-bytecode.texi [hunk 6]
```diff
@@ -1224,76 +1323,128 @@ Call @code{current-buffer} with no arguments.
...
+@unnumberedsubsubsec @code{save-current-buffer-1} (114)
+@kindex save-current-buffer
+Call @code{save-current-buffer} with one stack argument,
```
Comment (pipcet) [inline elisp-bytecode.texi]: There is no stack argument, as far as I can see.

File: elisp-bytecode.texi [hunk 7]
```diff
@@ -1331,9 +1482,145 @@ new binding when executed.  They need to be balanced with
...
+@item @code{concatN} (176)
+@item @code{insertN} (177)
+@end itemize
+
+Instruction @code{unbind-all} was added to support tail-recursion
+removal.  However this was never subsequently implemented so this
+intruction was never generated.
+
+Starting in this version, unless C prepocessor variable
+@code{BYTE_CODE_SAFE} (on by default) is defined, the obsolete
```
Comment (pipcet) [inline elisp-bytecode.texi]: off by default, I think.

File: elisp-bytecode.texi [hunk 8]
```diff
@@ -1331,9 +1482,145 @@ new binding when executed.  They need to be balanced with
...
+@node After 23.1 and Starting in 24.1
+@section After 23.1 and Starting in 24.1
+
+The following became obsolete:
+
+@unnumberedsubsec @code{save-current-buffer} (97)
+
+Is replaced by @code{save-current-buffer-1} (114)
+
+Unless C prepocessor variable @code{BYTE_CODE_SAFE} (on by default) is
```
Comment (pipcet) [inline elisp-bytecode.texi]: ditto

File: elisp-bytecode.texi [hunk 9]
```diff
@@ -1878,62 +2185,79 @@ Make a binding recording buffer, point, and mark.
 @item @verb{|0161|}
 @tab @verb{| 113|}
 @tab @verb{|  set-buffer|}
-@tab Call @code{set-buffer} with one argument.
-@tab @math{-1+1}
+@item @verb{|0162|}
+@tab @verb{| 114|}
+@tab @verb{|save-current-buffer-1|}
+@tab Call @code{save-current-buffer} with one argument.
```
Comment (pipcet) [inline elisp-bytecode.texi]: Again, I don't think this is best described as a "call" instruction.

File: elisp-bytecode.texi [hunk 10]
```diff
@@ -1878,62 +2185,79 @@ Make a binding recording buffer, point, and mark.
 @item @verb{|0161|}
 @tab @verb{| 113|}
 @tab @verb{|  set-buffer|}
-@tab Call @code{set-buffer} with one argument.
```
Comment (pipcet) [inline elisp-bytecode.texi:1881]: Why was this removed? Is it incorrect?

File: elisp-bytecode.texi [hunk 11]
```diff
@@ -896,6 +935,16 @@ Call @code{list} with three arguments.
 @kindex list4
 Call @code{list} with four arguments.
 
+@node listN
+@unnumberedsubsubsec @code{list4} (70)
+@kindex listN
+Call @code{list} an arbitrary number of arguments.
```
Comment (pipcet) [inline elisp-bytecode.texi]: Again, "up to 255".

**Review by pipcet — COMMENTED** [review]
Looks good overall! I'll be travelling for a few days, FWIW.

File: elisp-bytecode.texi (hunk 1 above)
Comment (rocky) [reply elisp-bytecode.texi:549]: If you have specific wording, let me know. I'm not sure what would change.

Turns out in 25.1 the behviour change prints an error and doesn't abort. I'll push the updates to that soon.

File: elisp-bytecode.texi (hunk 4 above)
Comment (rocky) [reply elisp-bytecode.texi]: That makes more sense of the generated code. Have fixed and in the opcode table it will be pushed as soon as I go over the rest.

File: elisp-bytecode.texi (hunk 11 above)
Comment (rocky) [reply elisp-bytecode.texi]: Thanks. Have fixed the xxxN, and will push when I am finished with the comments.

File: elisp-bytecode.texi (hunk 5 above)
Comment (rocky) [reply elisp-bytecode.texi]: Ok change, and will push soon.

File: elisp-bytecode.texi (hunk 8 above)
Comment (rocky) [reply elisp-bytecode.texi]: Ok. Changed and will go in next push.

File: elisp-bytecode.texi (hunk 9 above)
Comment (rocky) [reply elisp-bytecode.texi]: Ok. Suggest something.

File: elisp-bytecode.texi (hunk 10 above)
Comment (rocky) [reply elisp-bytecode.texi:1881]: Probably an inadvertent mistake. Have readded and will push soon.

File: elisp-bytecode.texi (hunk 6 above)
Comment (rocky) [reply elisp-bytecode.texi]: Thanks. I was wondering what was going on in the bytecode.

File: elisp-bytecode.texi [hunk 12]
```diff
@@ -896,6 +935,14 @@ Call @code{list} with three arguments.
 @kindex list4
 Call @code{list} with four arguments.
 
+@node listN
+@unnumberedsubsubsec @code{list4} (70)
```
Comment (pipcet) [inline elisp-bytecode.texi]: `listN` here

File: elisp-bytecode.texi (hunk 10 above)
Comment (rocky) [reply elisp-bytecode.texi:1881]: Addressed in 8a3903d7 . Not sure why this still shows. Also, I think we should remove the "with one argument" stuff as those are indicated in the stack effects. In this table, I think we want just the minimum and the full description in the Instructions section.  I would like a link with no extra verbiage which is why I started the other index. However I haven't been able to pull that off yet.

I think the stack effects might also appear the instruction section. However, I'll creating a couple more branches to try out some new ideas: 

* Breaking up the opcode table into a couple of sequential tables
* Better descriptions in the Instruction section.

Comment (pipcet) [reply elisp-bytecode.texi:1881]: I agree on th "with one argument" thing, it'd be best to omit both that and the descriptions and just write "call *" or "call +"

File: elisp-bytecode.texi (hunk 1 above)
Comment (pipcet) [reply elisp-bytecode.texi:549]: I'd just change "invalid bytecode and assembly instructions" to "invalid opcode".

File: elisp-bytecode.texi [hunk 13]
```diff
@@ -1133,6 +1244,7 @@ but simply call the corresponding C function.
 * bobp::
 * current-buffer::
 * set-buffer::
+* byte-save-current-buffer::
```
Comment (pipcet) [inline elisp-bytecode.texi:1247]: I'd mildly prefer `save-current-buffer-OBSOLETE` and `save-current-buffer` as our names for these instructions.

File: elisp-bytecode.texi [hunk 14]
```diff
@@ -1222,90 +1342,165 @@ Call @code{current-buffer} with no arguments.
 @node set-buffer
 @unnumberedsubsubsec @code{set-buffer} (113)
 @kindex set-buffer
-Call @code{set-buffer} with one argument.
+Call @code{set-buffer} with the top of the stack,
+The top of the stack is replaced by the result of
+@code{(set-buffer TOS)}
+
+@strong{Example}
+@code{(defun set-buffer-eg() (set-buffer "*scratch"))} generates:
```
Comment (pipcet) [inline elisp-bytecode.texi:1350]: I'd use "`*scratch*`" in this example.

File: elisp-bytecode.texi [hunk 15]
```diff
@@ -1222,90 +1342,165 @@ Call @code{current-buffer} with no arguments.
...
 @node discard
 @unnumberedsubsec @code{discard} (136)
 @kindex discard
-Discard one value.
+Discard one value from the stack.
+
+@node discardN
+@unnumberedsubsec @code{discardN} (180)
+@kindex discardN
+Discards up to 255 arguments from the stack.  Note there is a special
```
Comment (pipcet) [inline elisp-bytecode.texi:1500]: argument -> value, I'd say.

File: elisp-bytecode.texi [hunk 16]
```diff
@@ -1320,6 +1515,22 @@ PC  Byte  Instruction
...
+@unnumberedsubsec @code{stack-set} (180)
+@kindex stack-set
+Like discard. [What's the difference?[
+
+Added in Emacs 24.1
+
+@node stack-set2
+@unnumberedsubsec @code{stack-setN} (181)
+@kindex stack-set2
+Like discardN. [What's the difference?]
```
Comment (pipcet) [inline elisp-bytecode.texi:1528]: I don't think that's true. stack-set discards only one value, after copying it to a position further up the stack; however, stack-set[0] is indeed equivalent to discard.

File: elisp-bytecode.texi [hunk 17]
```diff
@@ -1320,6 +1515,22 @@ PC  Byte  Instruction
...
+Can handle up to 255 arguments.  Note there is a special
+instruction when there is only one argument.
```
Comment (pipcet) [inline elisp-bytecode.texi:1530]: stack-set handles up to 255 values, stack-set2 handles up to 65535 values.

File: elisp-bytecode.texi [hunk 18]
```diff
@@ -1331,9 +1542,272 @@ new binding when executed.  They need to be balanced with
...
+
+@node After 21.1 and Starting in 22.1
+@section After 21.1 and Starting in 22.1
+
+There were no instruction changes.
+
+The bytecode meta-comment no longer includess the bytecomp version used.
+
+Bytecode meta-comments look like this:
+@verbatim
```
Comment (pipcet) [inline elisp-bytecode.texi:1686]: what about the `;ELC^W^@^@^@` line? Is that part of the bytecode meta-comment?

File: elisp-bytecode.texi [hunk 19]
```diff
@@ -1344,164 +1818,164 @@ Make a binding recording buffer, point, and mark.
...
 @item @verb{|  05|}
 @tab @verb{|   5|}
-@tab @verb{|  stack-ref5|}
+@tab @verb{|  stack-ref[5]|}
 @tab stack reference 5
 @tab @math{+1}
 @item @verb{|  06|}
 @tab @verb{|   6|}
-@tab @verb{|  stack-ref6|}
+@tab @verb{|  stack-ref[6]|}
```
Comment (pipcet) [inline elisp-bytecode.texi:1850]: I think I messed up disassembling this instruction correctly, but I'll have to check. Should it be, say, `stack-ref[6] [255]`

File: elisp-bytecode.texi [hunk 20]
```diff
@@ -2244,7 +2828,7 @@ Make a binding recording buffer, point, and mark.
 @item @verb{|0326|}
 @tab @verb{| 214|}
 @tab @verb{|  constant-i22|}
-@tab
+     @tab
```
Comment (pipcet) [inline elisp-bytecode.texi:2831]: I don't understand this change, was it intentional?

**Review by pipcet — APPROVED** [review]
Let's merge this if it's any help at all, it's definitely a step in the right direction.
